### ISSUE TITLE
BUG: unread indicator count is not reset after logout reported by @parasharrajat

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#11671]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app on web.
2. Login with any account that has some unread messages.  (to get unread messages send some messages from a different account to the account being logged in).
3. Now wait a couple seconds.
4. Observe there are unread message count on the Tab.
5. Do not go to the chat that has unread messages.
6. Now Sign out.
7. Wait a couple seconds.
8. Unread count is shown back after log out.

Optional:
 9. If count is not shown by step 8, login back with the same account without refreshing the page.
 10. Wait for the count to show up.
 11. Log out now.
 12. Wait a couple seconds, observe the Unread count on the tab bar.

## Expected Result:
No unread count and indicator is shown after logout

## Actual Result:
Unread count and indicator are shown after logout on login page.

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Desktop App

**Version Number:** 1.2.12-2
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
https://user-images.githubusercontent.com/43996225/194615034-efb34a05-7dfd-4644-b813-d49dbc39827d.mp4

https://user-images.githubusercontent.com/43996225/194615559-d16582d9-ae1e-42e0-bd79-1e13f9e25b55.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @parasharrajat
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1665066610938259

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by tjferriss** [comment]
I'm going to pass this over to engineering as I think it's a pass. However, I am on the fence as to whether this issue really hits the _value_ part of the review:
>Value — in other words, why is this particular issue important? Are there reasonable workarounds? Is the value of resolution greater than the resources required to reach it?

This seems like a polish item and if it requires more than a super small amount of work then I'd suggest we pass for now.

**Comment by Justicea83** [comment]
This can be external, we set the unreadCount on the website [here](https://github.com/Expensify/App/blob/79d8bd205be0879d3914fe926eaced3629caa4a1/src/libs/UnreadIndicatorUpdater/updateUnread/index.website.js#L13).

**Comment by tienifr** [comment]
#Proposal

## Problem
Now, we're using `listenForReportChanges` function to listen the changing of the `reports` and then using function `throttledUpdatePageTitleAndUnreadCount` to update the pageTitle and unreadCount. In function `listenForReportChanges` we check `if (!report || !report.reportID)` we do nothing. That why when we logout we don't reset the reports.

##Solution

we'll create new function `resetReportAfterLogout` to reset the reports and use `UnreadIndicatorUpdater.resetReportAfterLogout()` in `componentDidMount` of `SignInPage`

In https://github.com/Expensify/App/blob/main/src/libs/UnreadIndicatorUpdater/index.js

```diff
+function resetReportAfterLogout(){
+    Object.keys(reports).forEach(k => delete reports[k])
+}

export default {
    listenForReportChanges,
    stopListeningForReportChanges,
    throttledUpdatePageTitleAndUnreadCount,
+    resetReportAfterLogout
};

```

In https://github.com/Expensify/App/blob/main/src/pages/signin/SignInPage.js#L48

```diff
-        setTimeout(() => updateUnread(0), 0);
+        setTimeout(() => {
+            UnreadIndicatorUpdater.resetReportAfterLogout()
+            updateUnread(0)
+        }, 0);
```

Uploading Screen Recording 2022-10-11 at 18.08.51.mp4…

**Comment by marcaaron** [comment]
@Justicea83 why did we unassign @CortneyOfstad?
@sobitneupane why did we add the `Help Wanted` label?

Did I miss a process change? Seems like we need to wait for the Upwork job before the Help Wanted label gets added (which should happen automatically when the `Exported` label is added).

**Comment by CortneyOfstad** [comment]
Sounds good — just about to head out for the day, but will tackle this ASAP in the morning to get the Upwork job created 👍

**Comment by Justicea83** [comment]
> @Justicea83 why did we unassign @CortneyOfstad? @sobitneupane why did we add the `Help Wanted` label?
> 
> Did I miss a process change? Seems like we need to wait for the Upwork job before the Help Wanted label gets added (which should happen automatically when the `Exported` label is added).

That must be a Github bug, I remember only unassigning myself 😁

**Comment by CortneyOfstad** [comment]
Posted job here — https://www.upwork.com/ab/applicants/1580165477000413184/job-details

**Comment by sobitneupane** [comment]
@marcaaron 
>Did I miss a process change?

Here is the [slack discussion](https://expensify.slack.com/archives/C02NK2DQWUX/p1663690490772079?thread_ts=1663665823.165569&cid=C02NK2DQWUX) regarding the process update.

**Comment by fedirjh** [comment]
# Proposal
## Problem
the `UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount()` is being invoked on `onStateChange`  on the `NavigationContainer` . the `throttledUpdatePageTitleAndUnreadCount` execute the `updateUnread` function after 100 ms . so when user logs out this happen: 
1. State Updated -> call to UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount()
2. SignInPage is mounted and `updateUnread` is executed immediately and update the title
3. `UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount()` , after 100 ms , execute `updateUnread` and update the title again 

## Solutions
### Solution 1
Delay the `updateUnread` on the SignInPage for 100 ms

https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/pages/signin/SignInPage.js#L48
```diff
-          setTimeout(() => updateUnread(0), 0);
+          setTimeout(() => updateUnread(0), 100);
```
### Solution 2 

Call `UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount()` only when user is authenticated

https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/Navigation/NavigationRoot.js#L35
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/Navigation/NavigationRoot.js#L49
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/Navigation/NavigationRoot.js#L62

 ```diff
-                parseAndLogRoute(state) {
+               parseAndLogRoute(state,authenticated) {

+               if (authenticated) {
                   UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount();
+               }
-                onStateChange={this.parseAndLogRoute}
+               onStateChange={(state) => this.parseAndLogRoute(state,this.props.authenticated)}

```

### Solution 3 

Remove Reports when `stopListeningForReportChanges` is executed 

https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/UnreadIndicatorUpdater/index.js#L7
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/UnreadIndicatorUpdater/index.js#L42-L48

```diff
-          const reports = {};
+         let reports = {};
...

function stopListeningForReportChanges() {
    if (!connectionID) {
        return;
    }
+       reports = {}

    Onyx.disconnect(connectionID);
}
```

**Comment by sobitneupane** [comment]
All solutions proposed by @fedirjh in [his proposal](https://github.com/Expensify/App/issues/11671#issuecomment-1276338610) solves this issue. I prefer the third solution because it also solves below mentioned issue.

While testing above proposals I found that after signing out from one account and signing in to another account, the unread indicator count from previously signed in account is displayed. It is displayed for few seconds if the unread indicator count for newly signed in account is non-zero and if the unread indicator count for newly signed in account is zero, it is displayed until the page is reload.

cc: @marcaaron 

🎀👀🎀 C+ reviewed

**Comment by tienifr** [comment]
@sobitneupane I think u miss somethings here, I retested my proposal and It works correctly. Maybe u haven't add the `import UnreadIndicatorUpdater from '../../UnreadIndicatorUpdater'` in https://github.com/Expensify/App/blob/main/src/pages/signin/SignInPage.js. If there is, please help recheck.
My evidences:

https://user-images.githubusercontent.com/113963320/195655273-3b74ea03-d000-46da-b00f-1c96da6efa0e.mp4

https://user-images.githubusercontent.com/113963320/195655306-8d90b795-53f9-4c8d-a90e-a0dab8f75a32.mp4


By the way, the third solution of @fedirjh seems to be flicked.


https://user-images.githubusercontent.com/113963320/195658147-e888b4f2-c718-4783-8382-a41093491fb2.mp4

**Comment by marcaaron** [comment]
Sorry I don't really like any of the solutions proposed by @fedirjh

**Solution 1** - setTimeout is hacky in this case as it's not clear what we are waiting for
**Solution 2** - It feels like the wrong place to apply this logic. Also we are passing a prop to a method that can access these props already because it's a class method.
**Solution 3** - It's not the job of that listener to remove the reports. This should happen when Onyx.clear() is called.

@sobitneupane I think we can wait for a higher quality solution. The one's I've seen are not really acceptable. Though the second solution related to authentication feels like the closest to what we might want - I think we'd want a different implementation.

**Comment by fedirjh** [comment]
@marcaaron thanks for the feedback , i have updated both solution 1 and 2 

for solution 1 : when `SignInPage` is mounted it execute `updateUnread(0)` immediately , but it should be deferred until `UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount();` is  . setting timeout to 1ms will resolve this issue 

```jsx
setTimeout(() => updateUnread(0), 1);
```
for solution 3 : the `reports`  const is used on the subscription callback , the job of `stopListeningForReportChanges`  is to remove the subscription callback , so the reports should be rest when we remove the subscription callback , if it's not reseted , the next login will use the old reports value . check demo

https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/UnreadIndicatorUpdater/index.js#L39-L42


[Screencast from 10-13-2022 08:25:35 PM.webm](https://user-images.githubusercontent.com/36869046/195691970-061a32ed-e99e-4789-be19-24842dad6b8b.webm)

**Comment by marcaaron** [comment]
>for solution 1 : when SignInPage is mounted it execute updateUnread(0) immediately , but it should be deferred until UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount(); is . setting timeout to 1ms will resolve this issue

We're still using a `setTimeout` so it's a no from me.

>for solution 3 : the reports const is used on the subscription callback , the job of stopListeningForReportChanges is to remove the subscription callback , so the reports should be rest when we remove the subscription callback , if it's not reseted , the next login will use the old reports value . check demo

I'm not seeing how this is a different solution from the one before. As mentioned already, the `reports` value should be reset when `Onyx.clear()` is called.

**Comment by fedirjh** [comment]
@marcaaron we are talking about reports in the **memory** 
```jsx 
const reports = {};
```
 not that one from **indexedDB**

**Comment by marcaaron** [comment]
>the actual code uses setTimeout , i just suggested a change of the delay

Apologies, in that case, I would entertain a proposal that removes the `setTimeout()`, but not one that modifies it without explaining why it's there to begin with.

>we are talking about reports in the memory

I understood this. Thanks.

**Comment by tienifr** [comment]
@marcaaron @sobitneupane What do you think about my proposal here https://github.com/Expensify/App/issues/11671#issuecomment-1274549195

**Comment by sobitneupane** [comment]
>What do you think about my proposal here [#11671 (comment)](https://github.com/Expensify/App/issues/11671#issuecomment-1274549195)

@tienifr Thank you for the proposal. But we are looking for proposals that removes the use of `setTimeout()`. And we believe there must be a way to tackle this issue without the use of `setTimeout()`

**Comment by fedirjh** [comment]
> While testing above proposals I found that after signing out from one account and signing in to another account, the unread indicator count from previously signed in account is displayed.

@sobitneupane this issue exists in staging & production , it displays the sum of unread reports from previous and active account , you can reproduce it by using two accounts whit unread count :

1. Open the app on web.
2. Login with any account that has some unread messages.
3. Observe there are unread message count on the Tab.
4. Do not go to the chat that has unread messages.
5. Now Sign out.
6. Login with another account that has some unread messages.
7. Now you see that the count is the sum of **unique** unread reports of previous and active account  

this is due the reports persists in memory , new logged in user's reports are being merged with previous logged in account 
as @marcaaron mentioned : 

>  the reports value should be reset when Onyx.clear() is called

here is solution :
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/UnreadIndicatorUpdater/index.js#L7

```jsx
-        const reports = {};
+        let reports = {}
+        function clearReports(){
+            reports = {}
+        }
...
export default {
    listenForReportChanges,
    stopListeningForReportChanges,
    throttledUpdatePageTitleAndUnreadCount,
+   clearReports
};
```
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/actions/SignInRedirect.js#L36-L38

```jsx
+   import UnreadIndicatorUpdater from "../UnreadIndicatorUpdater";
...

+   UnreadIndicatorUpdater.clearReports();
    Onyx.clear()
        .then(() => {
```

**Comment by marcaaron** [comment]
@fedirjh In the correct design the reports should be "cleared" already - there should be no reason to manually clear them before calling `Onyx.clear()`

**Comment by marcaaron** [comment]
@tienifr your solution has the same problem as @fedirjh - my expectation is that if we clear data with Onyx that any subscribers that set local values should have those values cleared automatically.

**Comment by fedirjh** [comment]
> my expectation is that if we clear data with Onyx that any subscribers that set local values should have those values cleared automatically.

@marcaaron thanks for clarification , the actual reports const is out of scope for the subscriber callback , then i would suggest those changes 
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/UnreadIndicatorUpdater/index.js#L14 https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/UnreadIndicatorUpdater/index.js#L25-L37 

```diff
-   const reports = {};
...
-   const throttledUpdatePageTitleAndUnreadCount = _.throttle(() => {
+   const throttledUpdatePageTitleAndUnreadCount = _.throttle((reports) => {
...
function listenForReportChanges() {
    connectionID = Onyx.connect({
        key: ONYXKEYS.COLLECTION.REPORT,
+        waitForCollectionCallback:true, 
-        callback: (report) => {
+        callback: (reports) => {
-            if (!report || !report.reportID) {
+            if (!reports) {
                return;
            }

-            reports[report.reportID] = report;
-            throttledUpdatePageTitleAndUnreadCount();
+            throttledUpdatePageTitleAndUnreadCount(_.filter(reports, null));
        },
    });
}
```
then no need to call `UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount();` in `NavigationRoot.js`
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/Navigation/NavigationRoot.js#L49
we just remove it
```dif
-   UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount();
```

For SignInPage we can remove the setTimeout , the setTimeout looks to fix a bug with chrome where title changes on the dom but title bar didn't update (check my screenshot) , for safari it's working fine 

https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/pages/signin/SignInPage.js#L48
```diff
componentDidMount() {
        // Always reset the unread counter to zero on this page
        // NOTE: We need to wait for the next tick to ensure that the unread indicator is updated
-        setTimeout(() => updateUnread(0), 0);
+        updateUnread(0)
    }
```

## Test

this is the chrome bug , the favicon updated fine but title only update on dom , the setTimeout looks to fix this bug
![Screenshot from 2022-10-14 19-35-52](https://user-images.githubusercontent.com/36869046/195917713-b89a8339-19ef-474c-9d6b-aed025c9f219.png)

**Comment by sobitneupane** [comment]
@fedirjh Why do we need to force update unread in signinpage?
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/pages/signin/SignInPage.js#L48

I believe it should be updated when onyx is cleared. We might have stopped listening for changes in report before Onyx is cleared.

**Comment by sobitneupane** [comment]
@fedirjh I agree with your reason below 
> this is due the reports persists in memory , new logged in user's reports are being merged with previous logged in account

Cause might be:
- reports is stored as global variable.
- When onyx is cleared and start listening for change in report (`listenForReportChanges()`), `!report` is `true` in callback function and the callback function returns. We are not clearing reports stored in global variable in any place.
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/UnreadIndicatorUpdater/index.js#L28-L36

**Comment by fedirjh** [comment]
@sobitneupane , @marcaaron When `Onyx.clear()` i called , there is a last time change to report where all reports are cleared (nulled) we need to wait for this change to happen , then we disconnect , if `Onyx.disconnect `is called immediately , it will be executed after `Onyx.clear()`  without waiting for that last change , in this case `setTimeout` will ensure enqueuing :

[Onyx.js#L908](https://github.com/Expensify/react-native-onyx/blob/4f2cd795499d3a94ffea78812a461c7baad38f55/lib/Onyx.js#L908)
```jsx
notifySubscribersOnNextTick(key, resetValue);
```

**Comment by fedirjh** [comment]
@sobitneupane i have updated my proposal , you can test it https://github.com/Expensify/App/issues/11671#issuecomment-1279341434

**Comment by marcaaron** [comment]
I'm going OOO for a week so we will need to find a new CME for this one. @sobitneupane please ping in Slack if you can.

**Comment by sobitneupane** [comment]
@fedirjh 
**Suggestion:** It is better to add updated proposal in new comment rather than editing previous comment. 

Rather than setTimeout, why not wait for onyx to be cleared before calling `stopListeningForReportChanges()`. 
> We might have stopped listening for changes in report before Onyx is cleared.

Can you please explain the line below?

> the actual reports const is out of scope for the subscriber callback 

Also, next point to consider is performance. We need to make our code better in performance? Your suggested changes does not look good performance wise? Because in each update You are suggesting to pass all the available reports to `throttledUpdatePageTitleAndUnreadCount` which is not the case now(only updated report are passed to `throttledUpdatePageTitleAndUnreadCount`). It might hinder performance when users have lot of reports.

**Comment by tienifr** [comment]
@sobitneupane what do you think about my updated proposal below:
In https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/UnreadIndicatorUpdater/index.js#L42
I will reset `reports` and then call function `throttledUpdatePageTitleAndUnreadCount` to update unread indicator

```diff
-    const reports = {};

```

```diff
function stopListeningForReportChanges() {
    if (!connectionID) {
        return;
    }
+    reports={}
+    throttledUpdatePageTitleAndUnreadCount();
    Onyx.disconnect(connectionID);
}
```

And then in https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/pages/signin/SignInPage.js#L45 
I can remove the logic using `setTimeout`

```diff
-    componentDidMount() {
-        // Always reset the unread counter to zero on this page
-        // NOTE: We need to wait for the next tick to ensure that the unread indicator is updated
-        setTimeout(() => updateUnread(0), 0);
-    }
```

**Comment by fedirjh** [comment]
> > the actual reports const is out of scope for the subscriber callback

@sobitneupane i meant that `reports` variable isn't declared inside the subscriber , so it persists in-memory 

> Why do we need to force update unread in signinpage ?

when Onyx is cleared , the subscriber callback doesn't update `reports` because it become null , `reports[report.reportID] = report;` here we add `report`  to `reports` variable , but when report become null (cleared) we can't remove it from reports as we don't have any reference to it , if Onyx.clear return the id instead of null then we can update the above code to remove report from reports , I can suggest a solution for this but it's not a good practice for performance , I think it's just better to force update count to zero `updateUnread(0)` 

here when report become null nothing happen :
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/UnreadIndicatorUpdater/index.js#L29-L31

## Proposal

- We have to declare `const reports = {};` inside the subscriber to ensure that data is cleared when subscriber is removed 
- When we stop stopListeningForReportChanges we just rest total count to zero 

### 1. **Changes in [UnreadIndicatorUpdater/index.js](https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/UnreadIndicatorUpdater/index.js#L42)**

```diff

-   const reports = {};
...
-   const throttledUpdatePageTitleAndUnreadCount = _.throttle(() => {
+   const throttledUpdatePageTitleAndUnreadCount = _.throttle((reports) => {
...
function listenForReportChanges() {
+   const reports = {};
    connectionID = Onyx.connect({
        key: ONYXKEYS.COLLECTION.REPORT,
        callback: (report) => {
           if (!report || !report.reportID) {
                return;
            }

            reports[report.reportID] = report;
            throttledUpdatePageTitleAndUnreadCount();
+           throttledUpdatePageTitleAndUnreadCount(reports);
        },
    });
}
...

function stopListeningForReportChanges() {
    if (!connectionID) {
        return;
    }
+   updateUnread(0);
    Onyx.disconnect(connectionID);
}
```
### 2. **Changes in [SignInRedirect.js ](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/actions/SignInRedirect.js#L53)**

To Avoid the use of setTimeout we should ensure that onyx.clear is executed then we execute stopListeningForReportChanges which will disconnect and reset counter to zero 
```diff
+        import UnreadIndicatorUpdater from '../UnreadIndicatorUpdater';

    Onyx.clear()
        .then(() => {
            ...
-        });
+        }).finally(() => UnreadIndicatorUpdater.stopListeningForReportChanges())
```
### 3. **Changes in [Session/index.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/actions/Session/index.js#L377)**
```diff
we have to remove `UnreadIndicatorUpdater.stopListeningForReportChanges()` from `cleanupSession`
function cleanupSession() {
    // We got signed out in this tab or another so clean up any subscriptions and timers
    NetworkConnection.stopListeningForReconnect();
-   UnreadIndicatorUpdater.stopListeningForReportChanges();
    PushNotification.deregister();
    PushNotification.clearNotifications();
    Pusher.disconnect();
    Timers.clearAll();
    Welcome.resetReadyCheck();
}
```
### 4. **Changes in [NavigationRoot.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/Navigation/NavigationRoot.js#L49)** 
there is no need for this line
```diff
-      UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount();
```
### 5. **Changes in [SignInPage.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/pages/signin/SignInPage.js#L45-L48)** 
there is no need anymore to force reset on this page , because we reset on 
```diff
class SignInPage extends Component {
-    componentDidMount() {
-        // Always reset the unread counter to zero on this page
-        // NOTE: We need to wait for the next tick to ensure that the unread indicator is updated
-        setTimeout(() => updateUnread(0), 0);
-    }
```

**Comment by sobitneupane** [comment]
@tjferriss As suggested earlier, we want `listenForReportChanges` to handle updating of the read/unread count. Since report is being changed, `throttledUpdatePageTitleAndUnreadCount();` should be called through `listenForReportChanges`.

**Comment by sobitneupane** [comment]
@fedirjh Your proposal steps 2-5 look convincing.

But in step 1, It is better to find a way to update unread `updateUnread` through `listenForReportChanges` >  `throttledUpdatePageTitleAndUnreadCount` > `updateUnread` if possible.

**Comment by fedirjh** [comment]
@sobitneupane here is an updated Proposal
## Proposal

### 1. **Changes in [UnreadIndicatorUpdater/index.js](https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/UnreadIndicatorUpdater/index.js#L42)**
we can get the `key` already , we can remove the `report` from `reports` when it's cleared 
```diff
function listenForReportChanges() {
    connectionID = Onyx.connect({
        key: ONYXKEYS.COLLECTION.REPORT,
-        callback: (report) => {
+        callback: (report,key) => {
-           if (!report || !report.reportID) {
+           if (!key) {
                return;
            }
+           if (!report) {
+               delete reports[key]
+            } else {
+                reports[key] = report;
+            }

-           reports[report.reportID] = report;
            throttledUpdatePageTitleAndUnreadCount();
        },
    });
}
```
### 2. **Changes in [SignInRedirect.js ](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/actions/SignInRedirect.js#L53)**

To Avoid the use of setTimeout we should ensure that onyx.clear is executed then we execute stopListeningForReportChanges which will disconnect and reset counter to zero 
```diff
+        import UnreadIndicatorUpdater from '../UnreadIndicatorUpdater';

    Onyx.clear()
        .then(() => {
            ...
-        });
+        }).finally(() => UnreadIndicatorUpdater.stopListeningForReportChanges())
```
### 3. **Changes in [Session/index.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/actions/Session/index.js#L377)**
```diff
we have to remove `UnreadIndicatorUpdater.stopListeningForReportChanges()` from `cleanupSession`
function cleanupSession() {
    // We got signed out in this tab or another so clean up any subscriptions and timers
    NetworkConnection.stopListeningForReconnect();
-   UnreadIndicatorUpdater.stopListeningForReportChanges();
    PushNotification.deregister();
    PushNotification.clearNotifications();
    Pusher.disconnect();
    Timers.clearAll();
    Welcome.resetReadyCheck();
}
```
### 4. **Changes in [NavigationRoot.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/Navigation/NavigationRoot.js#L49)** 
there is no need for this line
```diff
-      UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount();
```
### 5. **Changes in [SignInPage.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/pages/signin/SignInPage.js#L45-L48)** 
there is no need anymore to force reset on this page , because we reset on 
```diff
class SignInPage extends Component {
-    componentDidMount() {
-        // Always reset the unread counter to zero on this page
-        // NOTE: We need to wait for the next tick to ensure that the unread indicator is updated
-        setTimeout(() => updateUnread(0), 0);
-    }
```

**Comment by sobitneupane** [comment]
@fedirjh Your latest proposal looks promising. Have you tested it on other different scenarios where unread indicator is updated?

 @CortneyOfstad As Marc is OoO for a week (https://github.com/Expensify/App/issues/11671#issuecomment-1280072219), we need new CME here.

**Comment by fedirjh** [comment]
> Can you also clarify why UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount(); was needed before and how your proposed solution removes it's need?

@sobitneupane this change was made by @marcaaron https://github.com/Expensify/App/pull/10739/commits/211bd86236e5ccea86102cb25a8e452e86f5186a , i think the correct design is that title indicator is updated on report change only from the subscriber callback , not on state change , this will cause performance issue as it will fire on every state change (including route change)

**Comment by CortneyOfstad** [comment]
Since it has been over a week, doubled the price to $500. Also applying new CME now 👍

**Comment by CortneyOfstad** [comment]
Hmm, GH is being screwy with the labels. Trying that again!

**Comment by tsa321** [comment]
### Proposal

The problem is throttle update function callback is being called after user sign out and report memory is present. So we need to check if Onyx connection is present when updating total unread count. 

```diff
function stopListeningForReportChanges() {
    if (!connectionID) {
        return;
    }

      Onyx.disconnect(connectionID);
+   connectionID = undefined; 
}
```

and in throtled update function add check for connectionID:

```diff
const throttledUpdatePageTitleAndUnreadCount = _.throttle(() => {
-    const totalCount = _.filter(reports, ReportUtils.isUnread).length;
+    const totalCount = connectionID ? _.filter(reports, ReportUtils.isUnread).length : 0;
    updateUnread(totalCount);
}, 100, {leading: false});
```

**Comment by sobitneupane** [comment]
[Proposal](https://github.com/Expensify/App/issues/11671#issuecomment-1282419495) from @fedirjh looks good to me. I am not sure why `UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount();` was added in NavigationRoot.js. It was added in [this commit](https://github.com/Expensify/App/commit/211bd86236e5ccea86102cb25a8e452e86f5186a) by @marcaaron. @fedirjh is proposing to remove it. Changes test well.

cc: @puneetlath

**Comment by tsa321** [comment]
@sobitneupane  any advise for my [proposal](https://github.com/Expensify/App/issues/11671#issuecomment-1285811461)?

**Comment by sobitneupane** [comment]
> @sobitneupane any advise for my [proposal](https://github.com/Expensify/App/issues/11671#issuecomment-1285811461)? Is it bad? since the changes are minimal. .

@tsa321 Your proposal is similar to 3rd solution in [this proposal](https://github.com/Expensify/App/issues/11671#issuecomment-1276338610). You can find advise for your proposal in https://github.com/Expensify/App/issues/11671#issuecomment-1277944934 comment and comments following it.

**Comment by tsa321** [comment]
@sobitneupane 
Based on @marcaaron  advice:

> Solution 3 - It's not the job of that listener to remove the reports. This should happen when Onyx.clear() is called.

so we cannot remove reports variable inside `stopListeingForReportChanges`. 

How about:

```diff
.../UnreadIndicatorUpdater/index.js
+ let reports;
- const reports;
let connectionID;
....
const throttledUpdatePageTitleAndUnreadCount = _.throttle(() => {
-   //const totalCount = _.filter(reports, ReportUtils.isUnread).length;
+    const totalCount = connectionID? _.filter(reports, ReportUtils.isUnread).length : 0;
    updateUnread(totalCount);
}, 100, {leading: false});
....

function listenForReportChanges() {
    connectionID = Onyx.connect({
          key: ONYXKEYS.COLLECTION.REPORT,
+        waitForCollectionCallback: true,
+        callback: val => { 
+            reports = val;
+            throttledUpdatePageTitleAndUnreadCount();
+        }
-            if (!report || !report.reportID) {
-                return;
-           }
-            reports[report.reportID] = report;
-            throttledUpdatePageTitleAndUnreadCount();
-        },
    });
}
....
function stopListeningForReportChanges() {
    if (!connectionID) {
        return;
    }

      Onyx.disconnect(connectionID);
+    connectionID = undefined;
}
```

**Comment by sobitneupane** [comment]
Hello @tsa321, Thank you for your proposal. 

**Suggestion:** 
- At first please go through the discussion above. 
- Be sure your proposal is substantially different from other proposals. 

Your first proposal was same as 3rd solution of https://github.com/Expensify/App/issues/11671#issuecomment-1276338610  and your second proposal is almost similar to https://github.com/Expensify/App/issues/11671#issuecomment-1279341434.

Your 2nd proposal looks bit hacky. For further comments about your proposal please follow comments below https://github.com/Expensify/App/issues/11671#issuecomment-1279341434 proposal.

**Comment by tsa321** [comment]
Ok. I didn't see it. Thank you for your feedback.

**Comment by puneetlath** [comment]
Ok, let's go with @fedirjh's proposal. Let's make sure we do some very thorough testing on the PR to ensure we aren't introducing regressions on this one.

**Comment by fedirjh** [comment]
@puneetlath  @sobitneupane i have a better proposal . i have 2 points 

- We are already listening to report changes in the [Report.js](https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/actions/Report.js#L943) . isn't better to minimize the connections ? we can use that callback to watch the unread status , we don't have to create new connection 
- I don't recommend that we update the reports when we are clearing ( Onyx.clear) , i mean remove the report from the reports,   because if we have huge number of reports , all callback will be executed simultaneously this will make a delay before logging out . i tested with an account with 1.2k reports and there is a performance issue . So i suggest that we only listen for changes that are different from clearing , and finally reset the reports after Onyx.clear

https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/actions/Report.js#L923-L932

### 1. **Changes in [Report.js](https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/actions/Report.js#L923-L932)**
```diff
+ import * as ReportActionsUtils from '../ReportActionsUtils';
+ import UnreadIndicatorUpdater from '../UnreadIndicatorUpdater';

function handleReportChanged(report) { 
     if (!report) { 
         return; 
     } 
  
+     // Here we update the UnreadIndicatorUpdater from the handleReportChanged directly
+     if (ReportUtils.isUnread(report)) { 
+         UnreadIndicatorUpdater.addToUnread(report.reportID); 
+     } else { 
+         UnreadIndicatorUpdater.removeFromUnread(report.reportID); 
+     } 
```
### 2. **Changes in [UnreadIndicatorUpdater/index.js](https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/UnreadIndicatorUpdater/index.js#L42)**
we don't have to listen the report changes from here anymore , 
```diff
- const reports = {};
+ let reports = {};

...
const throttledUpdatePageTitleAndUnreadCount = _.throttle(() => {
-   const totalCount = _.filter(reports, ReportUtils.isUnread).length;
+   const totalCount = _.keys(reports).length;
    updateUnread(totalCount);
}, 100, {leading: false});

-function listenForReportChanges() {
-    connectionID = Onyx.connect({
-        key: ONYXKEYS.COLLECTION.REPORT,
-        callback: (report) => {
-            if (!report || !report.reportID) {
-                return;
-            }
-
-            reports[report.reportID] = report;
-            throttledUpdatePageTitleAndUnreadCount();
-        },
-    });
-}

-function stopListeningForReportChanges() {
-    if (!connectionID) {
-        return;
-    }
-
-    Onyx.disconnect(connectionID);
-}

+ /** 
+ * Add a report to the unread list 
+ * @param key 
+ */ 
+ function addToUnread(key) { 
+    reports[key] = true; 
+    throttledUpdatePageTitleAndUnreadCount(); 
+ } 
 
+ /** 
+ * Remove a report from the unread list 
+ * @param key 
+ */ 
+ function removeFromUnread(key) { 
+    if (!reports[key]) { 
+        return; 
+    } 
+    delete reports[key]; 
+    throttledUpdatePageTitleAndUnreadCount(); 
+ } 
 
+ /** 
+ * Clear the unread list and update the title and favicon 
+ */ 
+ function resetUnread() { 
+    reports = {}; 
+    throttledUpdatePageTitleAndUnreadCount(); 
+ } 
 
export default { 
-    listenForReportChanges, 
+    addToUnread, 
+    resetUnread, 
+    removeFromUnread, 
-    stopListeningForReportChanges, 
    throttledUpdatePageTitleAndUnreadCount, 
};

```
### 3. **Changes in [SignInRedirect.js ](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/actions/SignInRedirect.js#L53)**

To Avoid the use of setTimeout we should ensure that onyx.clear is executed then we execute resetUnread which will clear reports from memory and reset counter to zero 
```diff
+        import UnreadIndicatorUpdater from '../UnreadIndicatorUpdater';

    Onyx.clear()
        .then(() => {
            ...
-        });
+        }).finally(() => UnreadIndicatorUpdater.resetUnread())
```
### 4. **Changes in [Session/index.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/actions/Session/index.js#L377)**
we have to remove `UnreadIndicatorUpdater.stopListeningForReportChanges()` from `cleanupSession`
```diff
function cleanupSession() {
    // We got signed out in this tab or another so clean up any subscriptions and timers
    NetworkConnection.stopListeningForReconnect();
-   UnreadIndicatorUpdater.stopListeningForReportChanges();
    PushNotification.deregister();
    PushNotification.clearNotifications();
    Pusher.disconnect();
    Timers.clearAll();
    Welcome.resetReadyCheck();
}
```
### 5. **Changes in [Navigation/AppNavigator/AuthScreens.js](https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/Navigation/AppNavigator/AuthScreens.js#L105)**
we have to remove `UnreadIndicatorUpdater.listenForReportChanges()` from `AuthScreens` as we are not listening for changes from here
```diff
-       UnreadIndicatorUpdater.listenForReportChanges();
        App.openApp();
        App.setUpPoliciesAndNavigate(this.props.session);
        Timing.end(CONST.TIMING.HOMEPAGE_INITIAL_RENDER);
```
### 6. **Changes in [NavigationRoot.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/Navigation/NavigationRoot.js#L49)** 
there is no need for this line
```diff
-      UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount();
```
### 7. **Changes in [SignInPage.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/pages/signin/SignInPage.js#L45-L48)** 
there is no need anymore to force reset on this page , because we reset on Onyx.clear
```diff
class SignInPage extends Component {
-    componentDidMount() {
-        // Always reset the unread counter to zero on this page
-        // NOTE: We need to wait for the next tick to ensure that the unread indicator is updated
-        setTimeout(() => updateUnread(0), 0);
-    }
```

**Comment by fedirjh** [comment]
@sobitneupane  this is the issue when we have huge number of reports , title is updating multiple time during onyx clear , this is due to listening to clear changes , as I suggested , we don't have to listen to clear changes , and we have to reset reports after onyx.clear 


https://user-images.githubusercontent.com/36869046/197365023-c55b434d-0961-4e87-82b8-20ecbe0a1bff.mov

**Comment by sobitneupane** [comment]
@fedirjh What we are doing now is basically the same thing. Only difference is when user logout. To swiftly change unread reports to 0, we could have make use of `resetUnread()` of [the proposal](https://github.com/Expensify/App/issues/11671#issuecomment-1287935696). But it took around 20-30 seconds to change indicator to 0 in chrome.

**Comment by puneetlath** [comment]
@fedirjh can you estimate when the PR will be ready?

**Comment by sobitneupane** [comment]
@puneetlath We have [an issue](https://github.com/Expensify/App/issues/11671#issuecomment-1287939672) with the proposal. Expecting an updated proposal from @fedirjh.

**Comment by puneetlath** [comment]
Got it! @fedirjh when can we expect an updated proposal?

**Comment by fedirjh** [comment]
@sobitneupane [this issue](https://github.com/Expensify/App/issues/11671#issuecomment-1289926244) is related to the logic of `Onyx.clear` which i think it's not the right pattern , https://github.com/Expensify/react-native-onyx/blob/91aabb7219ddc8722cc2955de0398b1ae70c546d/lib/Onyx.js#L1030-L1040

```jsx
function clear() {
    return getAllKeys()
        .then((keys) => {
            _.each(keys, (key) => {
                const resetValue = lodashGet(defaultKeyStates, key, null);
                cache.set(key, resetValue);
                notifySubscribersOnNextTick(key, resetValue);
            });
            return Storage.clear();
        });
}
```
> But it took around 20-30 seconds to change indicator to 0 in chrome.
that's the time Onyx.clear took to clear and notify subscribers

@puneetlath if i will suggest an updated proposal i would suggest a change to the Onyx.clear logic . there are other bugs related to how Onyx.clear works like https://github.com/Expensify/App/issues/11901 and https://github.com/Expensify/App/issues/12055 , i have started a [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1667499042719369) on the slack channel

**Comment by sobitneupane** [comment]
@fedirjh Let's wait for views in [the thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1667499042719369). If we decide not to make change in Onyx, we can move forward with Marc's suggestion in the thread:
> Feels kind of edge case to me - but one way to solve could be some way to conditionally initialize Onyx with an empty reports list if you are not authenticated.

**Comment by fedirjh** [comment]
## Updated Proposal
hey @sobitneupane check this proposal : 

1. Add new key to onyx `UNREAD_REPORTS` that holds ids of unread reports
2. Push/remove id of report to/from `UNREAD_REPORTS` (we will use the connection inside the [Report.js](https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/actions/Report.js#L923-L932))
3. Watch `UNREAD_REPORTS` and update title

### Changes
#### Add new key to onyx
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/ONYXKEYS.js#L181-L182
```diff
    SHOULD_SHOW_COMPOSE_INPUT: 'shouldShowComposeInput',
+
+   // Holds ids of unread reports
+   UNREAD_REPORTS: 'unreadReports',
};
```
#### Push/remove id of report to/from `UNREAD_REPORTS`
https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/actions/Report.js#L923-L930
```diff
+ import * as ReportActionsUtils from '../ReportActionsUtils';
...
+ // Connect to unreadReports and init pushedToUnread
+ // We will use it check if report is pushed to unreadReports
+ let pushedToUnread = {};
+ Onyx.connect({
+    key: ONYXKEYS.UNREAD_REPORTS,
+    callback: unreadReports => pushedToUnread = unreadReports || {},
+ });

function handleReportChanged(report) {
    if (!report) {
        return;
    }

    if (report && report.reportID) {
+
+       // Check if the report is unread and if it is pushed to unreadReports
+       // Push it to unreadReports if it is unread and not already there
+       // This is to make sure that we don't push the same report to unreadReports multiple times
+       // Remove read reports from unreadReports if they are there
+       if ((!pushedToUnread[report.reportID] && ReportUtils.isUnread(report))
+           || (pushedToUnread[report.reportID] && !ReportUtils.isUnread(report))) {
+           Onyx.merge(ONYXKEYS.UNREAD_REPORTS, {[report.reportID]: ReportUtils.isUnread(report) || null});
+       }
```
#### Connect to  `UNREAD_REPORTS` and update the unread indicator
https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/Navigation/AppNavigator/AuthScreens.js#L65
```diff
+ // Connect to unreadReports and update the unread indicator
+ Onyx.connect({
+     key: ONYXKEYS.UNREAD_REPORTS,
+     callback: (unreadReports, key) => {
+         if (!key) {
+             return;
+         }
+         const unreadReportsCount = _.keys(unreadReports || {}).length;
+         UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount(unreadReportsCount);
+     },
+ });

const RootStack = createCustomModalStackNavigator();
...
-        // Listen for report changes and fetch some data we need on initialization
-        UnreadIndicatorUpdater.listenForReportChanges();
```

#### **Changes in [UnreadIndicatorUpdater/index.js](https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/UnreadIndicatorUpdater/index.js#L42)**
we don't have to listen the report changes from here anymore , 
```diff
- const reports = {};
...
- const throttledUpdatePageTitleAndUnreadCount = _.throttle(() => {
-   const totalCount = _.filter(reports, ReportUtils.isUnread).length;
-   const totalCount = _.keys(reports).length;
-    updateUnread(totalCount);
- }, 100, {leading: false});
+ const throttledUpdatePageTitleAndUnreadCount = _.throttle(updateUnread, 100, {leading: false});

-function listenForReportChanges() {
-    connectionID = Onyx.connect({
-        key: ONYXKEYS.COLLECTION.REPORT,
-        callback: (report) => {
-            if (!report || !report.reportID) {
-                return;
-            }
-
-            reports[report.reportID] = report;
-            throttledUpdatePageTitleAndUnreadCount();
-        },
-    });
-}

-function stopListeningForReportChanges() {
-    if (!connectionID) {
-        return;
-    }
-
-    Onyx.disconnect(connectionID);
-}
export default { 
-    listenForReportChanges, 
-    stopListeningForReportChanges, 
    throttledUpdatePageTitleAndUnreadCount, 
};

```
#### **Changes in [Session/index.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/actions/Session/index.js#L377)**
we have to remove `UnreadIndicatorUpdater.stopListeningForReportChanges()` from `cleanupSession`
```diff
function cleanupSession() {
-   UnreadIndicatorUpdater.stopListeningForReportChanges();
}
```
#### **Changes in [Navigation/AppNavigator/AuthScreens.js](https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/libs/Navigation/AppNavigator/AuthScreens.js#L105)**
we have to remove `UnreadIndicatorUpdater.listenForReportChanges()` from `AuthScreens` as we are not listening for changes from here
```diff
-       UnreadIndicatorUpdater.listenForReportChanges();
        App.openApp();
```
#### **Changes in [NavigationRoot.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/libs/Navigation/NavigationRoot.js#L49)** 
there is no need for this line
```diff
-      UnreadIndicatorUpdater.throttledUpdatePageTitleAndUnreadCount();
```
### 7. **Changes in [SignInPage.js](https://github.com/Expensify/App/blob/299f07bfa29c0dd81a6126691aae29e161097baa/src/pages/signin/SignInPage.js#L45-L48)** 
there is no need anymore to force reset on this page , because we reset on Onyx.clear
```diff
class SignInPage extends Component {
-    componentDidMount() {
-        // Always reset the unread counter to zero on this page
-        // NOTE: We need to wait for the next tick to ensure that the unread indicator is updated
-        setTimeout(() => updateUnread(0), 0);
-    }
```

**Comment by sobitneupane** [comment]
@fedirjh Let's wait for the slack thread. If you will be able to convince CME to go with your proposed idea in slack thread, we can go with it. Else, what about the following suggestion?
> @fedirjh Let's wait for views in [the thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1667499042719369). If we decide not to make change in Onyx, we can move forward with Marc's suggestion in the thread:
> > Feels kind of edge case to me - but one way to solve could be some way to conditionally initialize Onyx with an empty reports list if you are not authenticated.

**Comment by puneetlath** [comment]
@CortneyOfstad it feels like the current solution has stalled. Can you please double the price on this issue?

**Comment by CortneyOfstad** [comment]
Can do!

**Comment by CortneyOfstad** [comment]
Updated UpWork and shared in #Open-Source [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1668609123219779) 👍

**Comment by fedirjh** [comment]
@sobitneupane My last [updated proposal](https://github.com/Expensify/App/issues/11671#issuecomment-1305630104) fixes this [issue](https://github.com/Expensify/App/issues/11671#issuecomment-1287939672) , I don't know what we are waiting ? 

Oyx.clear delay / slow issue is a different issue which is present from time ago https://github.com/Expensify/App/pull/7397#discussion_r818023262 , and i think there is no intend to fix it as some issues that addresses it were closed : #12055 , #11901 , maybe it was an edge case ...

**Comment by sobitneupane** [comment]
> I don't know what we are waiting ?

@fedirjh You suggested it is related to onyx.clear(), so was waiting for the discussion to come to a conclusion. But looks like there is no progress in the discussion. 

@puneetlath 

- The issue with originally accepted [proposal](https://github.com/Expensify/App/issues/11671#issuecomment-1282419495) is: for large number of reports, unread account was not swiftly changing to 0.
https://user-images.githubusercontent.com/36869046/197365023-c55b434d-0961-4e87-82b8-20ecbe0a1bff.mov

- Issue on second [proposal](https://github.com/Expensify/App/issues/11671#issuecomment-1287935696): To reset report, it was taking around 20-30 seconds for account with large number of reports. Here @fedirjh suggested fix on `onyx.clear()` will fix the issue. Discussion [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1667499042719369). Not sure if it really is an issue.


Now we have two new proposals:
- @fedirjh has suggested add new key to onyx UNREAD_REPORTS to hold ids of unread reports [here](https://github.com/Expensify/App/issues/11671#issuecomment-1305630104)
- Marc [suggested](https://expensify.slack.com/archives/C01GTK53T8Q/p1667522742233869?thread_ts=1667499042.719369&cid=C01GTK53T8Q) to conditionally initialize Onyx with an empty reports if user not untenticated 
   > Feels kind of edge case to me - but one way to solve could be some way to conditionally initialize Onyx with an empty reports list if you are not authenticated.
   
   @puneetlath Which way will you prefer?

**Comment by marcaaron** [comment]
>Marc [suggested](https://expensify.slack.com/archives/C01GTK53T8Q/p1667522742233869?thread_ts=1667499042.719369&cid=C01GTK53T8Q) to conditionally initialize Onyx with an empty reports if user not untenticated

I want to clarify that, yes, that was a suggestion I made, but I think this conversation in general needs wider input from our internal engineering team.

**Comment by fedirjh** [comment]
@puneetlath , @sobitneupane , @marcaaron I have created the PR https://github.com/Expensify/App/pull/13035 for my proposal https://github.com/Expensify/App/issues/11671#issuecomment-1282419495

**Comment by sobitneupane** [comment]
@fedirjh Your proposal is not accepted yet. We are waiting feedback. So, won't be able to review the PR now.

**Comment by JmillsExpensify** [comment]
Just happening upon this issue, though I'm pretty unclear what the next steps are. It looks like we're at a bit of a standstill, is that right? Can we please create a new Slack thread or or use the existing one linked above to jumpstart next steps?

**Comment by JmillsExpensify** [comment]
Actually while I'm on this issue, my opinion is that we shouldn't fix this bug. We are deprecating unread counts in an upcoming release, are we not? Accordingly, I'm not really sure that there is a bug to fix, or that should be prioritized, because unread counts are going away entirely in the future.

**Comment by marcaaron** [comment]
I think post Unread Indicators doc we are still counting unread chats (just not chat message counts). Was there another conversation related to deprecating the counters entirely?

**Comment by trjExpensify** [comment]
Yeah, I don't recall that discussion. I remember the change from the unread indicators doc to move away from unread _messages_ to unread _chats_ instead, which I think is still where we're at as the intended design.

**Comment by puneetlath** [comment]
My apologies, I didn't realize this was waiting on me. I've updated the thread: https://expensify.slack.com/archives/C01GTK53T8Q/p1670946153675739?thread_ts=1667499042.719369&cid=C01GTK53T8Q

Also, @CortneyOfstad can you go ahead and update the job price in Upwork to $2000?

**Comment by CortneyOfstad** [comment]
Done and shared the update in the Open Source channel [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1670947299425099).

**Comment by getusha** [comment]
Still looking for new proposals?

**Comment by tgolen** [comment]
Since this is an ongoing issue with some problems getting proposal, I'm going to take this internal and work on a fix.

**Comment by puneetlath** [comment]
Sounds good Tim. If you want context on the discussion so far check out this thread: https://expensify.slack.com/archives/C01GTK53T8Q/p1670946153675739?thread_ts=1667499042.719369&cid=C01GTK53T8Q

**Comment by tgolen** [comment]
I've made progress here, but I've run into a bug in Onyx that I need to fix first.

When you subscribe to a collection in Onyx and use `waitForCollectionCallback` then call `Onyx.clear()` the subscriber callback is called n times with individual items. This isn't the correct behavior. The subscriber callback should only be called once with the entire collection.

**Comment by tgolen** [comment]
I opened a bug for ^ https://github.com/Expensify/App/issues/13584 and I am fixing it.

**Comment by JmillsExpensify** [comment]
> Yeah, I don't recall that discussion. I remember the change from the unread indicators doc to move away from unread messages to unread chats instead, which I think is still where we're at as the intended design.

Ah thanks, for setting me straight.

**Comment by JmillsExpensify** [comment]
Coming back to this issue though, let's like we're down to one PR, and working through one trailing bug/consideration before merging. It definitely looks close!

**Comment by tgolen** [comment]
This issue is going to require two different PRs to fix. The first PR is a refactoring of how the unread reports are counted. However, there was [one report](https://github.com/Expensify/App/pull/13600#issuecomment-1359997799) of Onyx data not fully clearing when the sign-out happens, which kept some unread reports sitting in Onyx and the unread count remained. 

Well, at least the count is accurately tracking what is in Onyx! 

I was not able to reproduce the issue of Onyx.clear() not working correctly, but if we can reproduce it reliably, and fix it, then this issue can be fully closed. There are quite a few changes being made to Onyx.clear() right now after switching over to SQLite and the `multiSet()` implementation being reverted, so it's a little murky on whether or not there is indeed a problem or not.

**Comment by fedirjh** [comment]
@tgolen i have reported that before https://github.com/Expensify/App/issues/11671#issuecomment-1302761477 ,  and i believe this issue was present from long time ago https://github.com/Expensify/App/pull/7397#discussion_r818023262

> @puneetlath if i will suggest an updated proposal i would suggest a change to the Onyx.clear logic . there are other bugs related to how Onyx.clear works like https://github.com/Expensify/App/issues/11901 and https://github.com/Expensify/App/issues/12055 , i have started a [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1667499042719369) on the slack channel

**Comment by tgolen** [comment]
@fedirjh Thanks for those reminders. I think what I would like to do is open a new issue that specifically tracks the problems with Onyx.clear() and not just the symptoms of it. I'll work on creating that GH and I will CC you on the issue.

If you have the time, I'd appreciate your review and help with testing https://github.com/Expensify/react-native-onyx/pull/219

**Comment by tgolen** [comment]
I've opened https://github.com/Expensify/App/issues/13884 to address issues related to Onyx.clear() not fully clearing out data

**Comment by fedirjh** [comment]
@tgolen thank for creating new I issue for Onyx.clear() , unfortunately that's conversation (#13884) is locked and I can't reply there.

**Comment by tgolen** [comment]
OK, I've unlocked it for now.

**Comment by sobitneupane** [comment]
PR deployed to production
<img width="1111" alt="Screenshot 2023-01-10 at 00 16 57" src="https://user-images.githubusercontent.com/25876548/211381707-71947594-38fd-4bb2-a3aa-d61957785428.png">

**Comment by sobitneupane** [comment]
@CortneyOfstad This should be ready for payment. [Associated PR](https://github.com/Expensify/App/pull/13600) was deployed to production last week.
<img width="986" alt="Screenshot 2023-01-12 at 21 07 04" src="https://user-images.githubusercontent.com/25876548/212107060-4df2eb5f-5b19-498c-9936-50cd96ef9105.png">

**Comment by CortneyOfstad** [comment]
@sobitneupane I am so sorry for the delay on this! For the payment, it would be like this, correct?

@parasharrajat gets $250 for reporting
@sobitneupane and @fedirjh each get $2000 for contributing and proposing the solution. 

Are there any bonuses that I am missing?

TIA!

**Comment by sobitneupane** [comment]
Actually, we end up solving this issue internally ([PR](https://github.com/Expensify/App/pull/13600)). So, Maybe

Rajat $250 for reporting.
Sobit $1000 for Internal PR review.

**Comment by CortneyOfstad** [comment]
Oh okay, thank you for the clarification! I'll get the contracts sent over in UpWork ASAP to have them accepted and I'll get payment sent out right away this morning. Again, I apologize for the delay on this!!

**Comment by CortneyOfstad** [comment]
Alright, contracts have been sent out @parasharrajat and @sobitneupane! Thanks!

**Comment by sobitneupane** [comment]
> Again, I apologize for the delay on this!!

No Problem @CortneyOfstad.

> Alright, contracts have been sent out

Accepted the offer.

**Comment by CortneyOfstad** [comment]
^^ @sobitneupane can you confirm?

**Comment by CortneyOfstad** [comment]
Okay, let me check with the team on this. Thank you!

**Comment by CortneyOfstad** [comment]
Thanks for your patience! After reviewing with the team, it was determined that this does not qualify for payment.

**Comment by CortneyOfstad** [comment]
Actually @fedirjh I misspoke!

**Comment by CortneyOfstad** [comment]
Our team has decided that this **does** qualify for payment and you'll get $250. I'll get that contract over to you ASAP!

**Comment by CortneyOfstad** [comment]
@fedirjh Contract sent over!

**Comment by CortneyOfstad** [comment]
Paid! Thank you!
